include/rocksdb/trace_record.h

```diff
@@ -0,0 +1,232 @@
+
+class ColumnFamilyHandle;
+class DB;
+class Status;
+
+// Supported trace record types.
+enum TraceType : char {
+  kTraceNone = 0,
+  kTraceBegin = 1,
+  kTraceEnd = 2,
+  // Query level tracing related trace types.
+  kTraceWrite = 3,
+  kTraceGet = 4,
+  kTraceIteratorSeek = 5,
+  kTraceIteratorSeekForPrev = 6,
+  // Block cache tracing related trace types.
+  kBlockTraceIndexBlock = 7,
+  kBlockTraceFilterBlock = 8,
+  kBlockTraceDataBlock = 9,
+  kBlockTraceUncompressionDictBlock = 10,
+  kBlockTraceRangeDeletionBlock = 11,
+  // IO tracing related trace type.
+  kIOTracer = 12,
+  // Query level tracing related trace type.
+  kTraceMultiGet = 13,
+  // All trace types should be added before kTraceMax
+  kTraceMax,
+};
+
+class WriteQueryTraceRecord;
+class GetQueryTraceRecord;
+class IteratorSeekQueryTraceRecord;
+class MultiGetQueryTraceRecord;
+
+// Base class for all types of trace records.
+class TraceRecord {
+ public:
+  TraceRecord();
+  explicit TraceRecord(uint64_t timestamp);
+  virtual ~TraceRecord();
+
+  virtual TraceType type() const = 0;
+
+  virtual uint64_t timestamp() const;
+
+  class Handler {
+   public:
+    virtual ~Handler() {}
+
+    virtual Status Handle(const WriteQueryTraceRecord& record) = 0;
+    virtual Status Handle(const GetQueryTraceRecord& record) = 0;
+    virtual Status Handle(const IteratorSeekQueryTraceRecord& record) = 0;
+    virtual Status Handle(const MultiGetQueryTraceRecord& record) = 0;
+  };
+
+  class ExecutionHandler : public Handler {
+   public:
+    ExecutionHandler(DB* db, const std::vector<ColumnFamilyHandle*>& handles);
+
+    ExecutionHandler(
+        DB* db,
+        const std::unordered_map<uint32_t, ColumnFamilyHandle*>& cf_map);
+
+    virtual ~ExecutionHandler() override;
+
+    virtual Status Handle(const WriteQueryTraceRecord& record) override;
+    virtual Status Handle(const GetQueryTraceRecord& record) override;
+    virtual Status Handle(const IteratorSeekQueryTraceRecord& record) override;
+    virtual Status Handle(const MultiGetQueryTraceRecord& record) override;
+
+   private:
+    DB* db_;
+    std::unordered_map<uint32_t, ColumnFamilyHandle*> cf_map_;
+  };
+
+  virtual Status Accept(Handler* handler) = 0;
+
+ private:
+  // Timestamp (in microseconds) of this trace.
+  uint64_t timestamp_;
+};
+
+// Base class for all query types of trace records.
+class QueryTraceRecord : public TraceRecord {
+ public:
+  explicit QueryTraceRecord(uint64_t timestamp);
+
+  virtual ~QueryTraceRecord() override;
+};
+
+// Trace record for DB::Write() operation.
+class WriteQueryTraceRecord : public QueryTraceRecord {
+ public:
+  WriteQueryTraceRecord(const WriteBatch& batch, uint64_t timestamp);
+
+  WriteQueryTraceRecord(WriteBatch&& batch, uint64_t timestamp);
+
+  WriteQueryTraceRecord(const std::string& rep, uint64_t timestamp);
+
+  WriteQueryTraceRecord(std::string&& rep, uint64_t timestamp);
+
+  WriteQueryTraceRecord(std::shared_ptr<WriteBatch> batch, uint64_t timestamp);
```

Comment:
I think we should just do `WriteQueryTraceRecord(PinnableSlice&& batch_rep, uint64_t timestamp)` and `Slice batch_rep()` now and worry about how to make a `WriteBatch` without copy later. I'm sure we can figure out a way to make a `WriteBatch` that reads from a `Slice`.

Reply:
Done